lib/Dialect/FIRRTL/Transforms/Lint.cpp

```diff
@@ -83,18 +102,67 @@ class Linter {
 
     return success();
   }
+
+  LogicalResult checkXmr(XMRDerefOp op) {
+    // XMRs under layers are okay.
+    if (op->getParentOfType<LayerBlockOp>() ||
+        op->getParentOfType<sv::IfDefOp>())
+      return success();
+
+    // The XMR is not under a layer.  This module must never be instantiated in
+    // the design.  Intentionally do NOT use "effective" design as this could
+    // lead to false positives.
+    if (!instanceInfo.allInstancesInDesign(fModule))
```

Comment:
I think this should be "not any". It would be nice if we could move this check out of the tight loop.

Here is an example of a missed XMR:
```mlir

firrtl.circuit "TB" {
  hw.hierpath private @xmrPath [@Foo::@sym]

  firrtl.module @Foo() {
    %a = firrtl.wire sym @sym : !firrtl.uint<1>
    %0 = firrtl.xmr.deref @xmrPath : !firrtl.uint<1>
  }

  firrtl.module @TB() {
    firrtl.instance dut @DUT()
    firrtl.instance foo @Foo()
  }

  firrtl.module @DUT() attributes {
    annotations = [
      {class = "sifive.enterprise.firrtl.MarkDUTAnnotation"}
    ]
  } {
    firrtl.instance foo @Foo()
  }
}
```

Reply:
Good catch! This is fixed in the forced update to: https://github.com/llvm/circt/pull/8668/commits/b5fd923a6755190d5f9b86dcdd583492244b7303

Reply:
Also, nice test. I included a distilled version of this in: 7e6892a49